src/libgap-api.h

```diff
@@ -6,6 +6,86 @@
 #include "system.h"
 
 
+#ifdef __GNUC__
+#define unlikely(x) __builtin_expect(!!(x), 0)
+#else
+#define unlikely(x) (x)
+#endif
+
+
+#ifndef GAP_ENTER_DEBUG
+#define GAP_ENTER_DEBUG 0
+#endif
+
+
+extern syJmp_buf * GAP_GetReadJmpError(void);
+extern void GAP_EnterDebugMessage_(char * message, char * file, int line);
+extern void GAP_EnterStack_(void *);
+extern void GAP_LeaveStack_(void);
+extern int GAP_Error_Prejmp_(const char *, int);
+extern int GAP_Error_Postjmp_(int);
+
+
+#if GAP_ENTER_DEBUG
+#define GAP_ENTER_DEBUG_MESSAGE(message, file, line)                         \
+    GAP_EnterDebugMessage_(message, file, line)
+#else
+#define GAP_ENTER_DEBUG_MESSAGE(message, file, line)                         \
+    do {                                                                     \
+    } while (0)
+#endif
+
+
+#ifdef __GNUC__
+#define GAP_EnterStack()                                                     \
+    do {                                                                     \
+        GAP_ENTER_DEBUG_MESSAGE("EnterStack", __FILE__, __LINE__);           \
+        GAP_EnterStack_(__builtin_frame_address(0));                         \
+    } while (0)
+#else
+// Generic fallback, should work in most cases (see e.g. GenStackFuncBags
+// in gasman.c)
+#define GAP_EnterStack()                                                     \
+    do {                                                                     \
+        GAP_ENTER_DEBUG_MESSAGE("EnterStack", __FILE__, __LINE__);           \
+        int _gap_dummy[1];                                                   \
+        GAP_EnterStack(_gap_dummy);
+    } while (0)
+#endif
+
+#define GAP_LeaveStack()                                                     \
+    GAP_LeaveStack_();
+
+
+#define GAP_Error_Setjmp()                                                   \
+    (GAP_Error_Prejmp_(__FILE__, __LINE__) ||                                \
+     GAP_Error_Postjmp_(sySetjmp(*GAP_GetReadJmpError())))
+
+
+/* Code which uses the GAP API--in particular any code in which GAP objects
+ * are
+ * allocated in local variables--must begin with this macro, and with the
+ * GAP_Leave() macro.
+ *
+ * GAP_Enter() is a combination of GAP_Error_Setjmp() and GAP_EnterStack().
+ * These may be used separately as well, but for most code that uses the
+ * GAP API you will want to use both anyways.  It is deliberately written
+ * with GAP_Error_Setjmp() first so that it can be used like
```

Comment:
I guess my problem here is just that I don't know *who* this inline documentation is for, since it doesn't get compiled into any prose or API documentation.  It would be better if there were some documentation of the GAP API in the GAP docs themselves: I hate having to read headers to get documentation (though I'm glad when they do have some).  Your suggested phrasing is correct though, so if you think it's clearer I'll use it.

Your suggestion about how to possibly rewrite the macro wouldn't work.  Part of the technical motivation for writing it the way it is (albeit not explicitly stated) is to do with how Cython does exception handling.  When used in Cython code, we want to mark `GAP_Enter()` as a "function" that can cause a Python exception to be raised.  I don't know how familiar you are with Cython, so forgive me if I'm repeating what you already know, but in Cython you can [declare](https://cython.readthedocs.io/en/latest/src/userguide/language_basics.html#error-return-values) that a C function (or macro--it doesn't make a distinction for these purposes) "raises an exception" based on its return value.  For example, if I declare:

```
cdef int GAP_Enter() except 0
```

that means that if `GAP_Enter()` returns `0`, Cython should assume that a Python-level exception occurred, and should jump to the exception handling code.  Therefore a call of `GAP_Enter()` in Cython code generates C code that looks roughly like in the example from the comment:

```
int ok = GAP_Enter(); if (unlikely(ok == 0) goto python_exception_handling;
```

So this was really written to this specific case.  However, making it any more complicated further narrows is applicability.  If someone else wanted to use it like you suggested in their own C code they could easily write their own wrappers around this simpler definition.

Reply:
I know nothing about Cython, so that explanation was helpful. And actually, I think it would be helpful if it was in the comment or source file.

For the time being, there simply is no reference manual for this GAP API, and unless somebody contributes one (in a format that works for us, etc. etc., so a random code drop isn't going to settle this), I am afraid there won't be one, ever. So all that people have is that header file, and hence it should at least explain the basics for users. I fully agree that this is bad, but it simply is the reality.

Moreover, obviously GAP developers also will have to rely on these comments. And any explanations in here will help preserve desired properties across refactorings of the code. Any such properties which are not documented or tested.

Anyway, the consequence is that header comments should ideally try to cater to both groups. But it's still better to address just users or just developers, than to write a mish-mash which helps neither ;-).

Reply:
Anyway, thanks for the updates, I'll have a look at them shortly (and I would wish that I wasn't the only one to look at this stuff...  *sigh* Also, I don't want claim to be an all-knowing authority on everything (that'd obviously be idiotic nonsense)! these are just my thoughts, and I'd encourage you to argue with anything I write that you disagree with.

Reply:
> So all that people have is that header file, and hence it should at least explain the basics for users. I fully agree that this is bad, but it simply is the reality.

I understand.  If I had more time I would happily contribute a reference manual.  I will certainly consider working on it: If at least the header is well documented it shouldn't be too difficult to convert that documentation to a manual.  In the meantime I understand we're all stretched too thin as it is, so I'll keep in mind that this header should also serve as good documentation for users.

> and I would wish that I wasn't the only one to look at this stuff

Aye, I know the feeling.